include/unifex/task.hpp

```diff
@@ -124,25 +125,25 @@ struct _task<T>::type {
     }
 
     template(typename Value)
-      (requires callable<decltype(unifex::await_transform), promise_type&, Value>)
+        (requires callable<decltype(unifex::await_transform), type&, Value>)
     auto await_transform(Value&& value)
-        noexcept(is_nothrow_callable_v<decltype(unifex::await_transform), promise_type&, Value>)
-        -> callable_result_t<decltype(unifex::await_transform), promise_type&, Value> {
+        noexcept(is_nothrow_callable_v<decltype(unifex::await_transform), type&, Value>)
+        -> callable_result_t<decltype(unifex::await_transform), type&, Value> {
       return unifex::await_transform(*this, (Value&&)value);
     }
 
     template(typename Value)
-        (requires convertible_to<Value, T>)
+        (requires constructible_from<T, Value>)
```

Comment:
Why was this changed?
The use of `convertible_to` was intended to only allow returning things that were implicitly convertible to the return-type here (same as for normal functions).

Reply:
This function constructs a `T` from a `Value`, so `constructible_from<T, Value>` seems right to me. I think you're saying you want to constrain it _as if_ the continuation were a function that takes `T` and we're calling it with a `Value`. But that's not the reality.

Reply:
Say we have the following:
```
struct X {};
struct Y {
  explicit Y(X);
};
```
then with an ordinary function I will get a compile-error if I try to do:
```
Y f() {
  return X{}; // error: no implicit conversion from X to Y
  return {X{}}; // error: constructor is explicit need to name the type Y
  return Y{X{}}; // OK
}
```
and the idea is to do the same for coroutines:
```
task<Y> f() {
  co_return X{}; // error: no implicit conversion from X to Y
  co_return {X{}}; // error: Y constructor is explicit
  co_return Y{X{}}; // OK
}
```
Although with the current definition of `return_value()` it won't allow the second form.
(We need to add a non-template overload for this to work).

My understanding is that with your change here to use `constructible_from` instead of `convertible_to` is that this will allow the first form to be allowed (`Y` is indeed `constructible_from` `X`).

Do we need to require both `convertible_to` and `constructible_from` then?

I've just used `convertible_to` on its own in other places where I want to require implicit convertibility and if we do need to also use `constructible_from` then I'd like to understand better the ways in which `convertible_to` is insufficient.

Reply:
Just because a type is convertible from another doesn't mean it is constructible from the other.

```C++
struct Y;
struct X {
    X() = default;
    explicit X(Y) = delete;
};
struct Y {
    operator X() const { return {}; }
};

static_assert(std::is_convertible_v<Y, X>);
static_assert(!std::is_constructible_v<X, Y>);
```

We should be requiring both.